### PR TITLE
import partr code, allow using it in threaded loops

### PR DESCRIPTION
**Pull request by JeffBezanson**
Rebased version of #30731.

### CODE REVIEW

**Comment by vchuravy** [comment]
So if I read this right tasks are only processed on the workers if we are in a threaded loop, but only the ones that are started by `@threads` and not the ones that are queued up outside?

**Comment by vtjnash** [comment]
It'll make progress on all tasks while inside an `@threads` block. Outside that block, the logic to automatically (dis)engage the `@threads` logic is currently not implemented.

File: base/task.jl [hunk 1]
```diff
@@ -446,10 +446,11 @@ function enq_work(t::Task)
     (t.state == :runnable && t.queue === nothing) || error("schedule: Task not runnable")
     tid = (t.sticky ? Threads.threadid(t) : 0)
```
Comment (vtjnash) [inline base/task.jl]: I think a better version of this function would be:
```suggestion
if t.sticky
    tid = Threads.threadid(t)
    if tid == 0
        tid = Threads.threadid()
    end
    push!(Workqueues[tid], t)
else
    tid = 0
    ccall(:jl_enqueue_task, Cvoid, (Any,), t)
end
```

File: src/task.c [hunk 2]
```diff
@@ -495,6 +495,7 @@ JL_DLLEXPORT jl_task_t *jl_new_task(jl_function_t *start, jl_value_t *completion
     t->excstack = NULL;
     t->stkbuf = NULL;
     t->started = 0;
+    t->prio = -1;
```
Comment (vtjnash) [inline src/task.c]:
```suggestion
    t->prio = -1;
    t->tid = 0; (and remove comment about "allow non-sticky tasks")
```

**Comment by AzamatB** [comment]
Does this mean that the functionality of [#22631](https://github.com/JuliaLang/julia/pull/22631) is now added to the Julia?

**Pull request merged**